tests/vm_state_test.go

```diff
@@ -3,40 +3,69 @@ package tests_test
 import (
 	"context"
 	"fmt"
-	"time"
-
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/util/rand"
+	"kubevirt.io/kubevirt/tests/libdv"
 	"kubevirt.io/kubevirt/tests/testsuite"
-
-	"kubevirt.io/kubevirt/tests/libnet"
-
-	"kubevirt.io/kubevirt/tests/libvmi"
+	"time"
 
 	expect "github.com/google/goexpect"
 	. "github.com/onsi/ginkgo/v2"
 	. "github.com/onsi/gomega"
+	k8sv1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/api/errors"
 	k8smetav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/utils/pointer"
 	v1 "kubevirt.io/api/core/v1"
 	"kubevirt.io/client-go/kubecli"
-
+	cdiv1 "kubevirt.io/containerized-data-importer-api/pkg/apis/core/v1beta1"
 	backendstorage "kubevirt.io/kubevirt/pkg/storage/backend-storage"
 	"kubevirt.io/kubevirt/tests"
 	"kubevirt.io/kubevirt/tests/console"
 	"kubevirt.io/kubevirt/tests/decorators"
+	"kubevirt.io/kubevirt/tests/framework/checks"
+	"kubevirt.io/kubevirt/tests/libnet"
+	"kubevirt.io/kubevirt/tests/libstorage"
+	"kubevirt.io/kubevirt/tests/libvmi"
 	"kubevirt.io/kubevirt/tests/libwait"
+	"kubevirt.io/kubevirt/tests/util"
 )
 
-var _ = Describe("[sig-storage]VM state", decorators.SigStorage, decorators.RequiresRWXFilesystemStorage, func() {
+type vmStatePersistenceOption struct {
+	WithTPM           bool
+	WithEFI           bool
+	BlockStorageClass *bool
+	VMDisk            *string
+	OrderedOperations []string
+}
+
+var (
+	vmStateBlockDVDisk       = "dv-block-order-1"
+	vmStateFilesystemPVCDisk = "pvc-fs-order"
+)
+
+var _ = Describe("[sig-storage]VM state", decorators.SigStorage, func() {
 	var virtClient kubecli.KubevirtClient
 	var err error
+	var dv *cdiv1.DataVolume
+	var pvc *k8sv1.PersistentVolumeClaim
 
 	BeforeEach(func() {
 		virtClient, err = kubecli.GetKubevirtClient()
 		Expect(err).ToNot(HaveOccurred())
+
+		// We always create the DV and PVC with small sizes no matter if they are
+		// used or not. This will simplify the code.
+		dv, pvc, err = prepareVMStateDVAndPVC(virtClient, util.NamespaceTestDefault)
```

Comment:
But then we are not testing auto-creation anymore, no? I'm confused by this line.

Reply:
So we added a sub-feature that if the `VMStateStorageClass` or `VMStateVolumeMode` is not specified in the kubevirt CR, virt-controller will determine them from the VM's DV/PVC disk. Specifically, virt-controller will iterate over all the disks of the VM based on their boot order, and use the storage class and the volume mode of the first seen DV/PVC disk to create the VM state PVC.

Here in the test, we attach a block DV and a filesystem PVC to the VM to test this behavior.